src/hotspot/share/gc/shenandoah/shenandoahVerifier.cpp

```diff
@@ -1333,12 +1333,40 @@ void ShenandoahVerifier::help_verify_region_rem_set(ShenandoahHeapRegion* r, She
       } else {
         // This object is not live so we don't verify dirty cards contained therein
         HeapWord* tams = ctx->top_at_mark_start(r);
+        HeapWord* orig_obj_addr = obj_addr;
         obj_addr = ctx->get_next_marked_addr(obj_addr, tams);
+        confirm_filled(orig_obj_addr, obj_addr, message);
       }
     }
   }
 }
 
+void ShenandoahVerifier::confirm_filled(HeapWord* start, HeapWord* end, const char *message) {
+  ShenandoahGenerationalHeap* gen_heap = ShenandoahGenerationalHeap::heap();
+  assert(gen_heap->mode()->is_generational(), "Filled is for generational");
+  ShenandoahOldGeneration* old_gen = gen_heap->old_generation();
+  ShenandoahOldHeuristics* old_heuristics = (ShenandoahOldHeuristics*) old_gen->heuristics();
+
+  bool check_fill_objects = !old_heuristics->has_coalesce_and_fill_candidates();
+  while (start < end) {
+    oop obj = cast_to_oop(start);
+
+    // This object is not marked.  It should be a fill object, or an object that is about to be filled.  If it is a fill
+    // object, it should be array of int.
+
+    // if there are coalesce-and-fill candidates, skip over this object.  Otherwise, this regions has already been
+    // coalesced and filled, so this should be a fill object.
+    if (check_fill_objects && !obj->is_array()) {
+      log_info(gc)("confirm_filled thinks thinks is_old_bitmap_stable(): %s, old_gen->is_mark_complete(): %s",
+                   gen_heap->is_old_bitmap_stable()? "yes": "no", old_gen->is_mark_complete()? "yes": "no");
+      ShenandoahAsserts::print_failure(ShenandoahAsserts::_safe_all, obj, start, nullptr, message,
```

Comment:
Why not `assert` here?

Reply:
Asserts:print_failure() makes it easier to provide more diagnostic information than would be provided by an assert failures.